File: Data Structures & Algorithms/max-water-container/submission-4.cpp
class Solution {
public:
    int maxArea(vector<int>& heights) {
        int l = 0, r = heights.size()-1;
        int res = 0;
        while(l < r){
            int size = min(heights[l], heights[r]) * (r-l);
            res = max(size, res);
            if(heights[l]<heights[r]){
                l++;
            }else{
                r--;
            }
        }
        return res;
    }
};

File: Data Structures & Algorithms/maximum-product-subarray/submission-4.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        curmin, curmax = 1, 1
        res = max(nums)
        for i in range(len(nums)):
            if nums[i] == 0:
                curmin, curmax = 1,1
                continue
            tmp = curmax
            curmax = max(curmax* nums[i], curmin * nums[i], nums[i])
            curmin = min(tmp* nums[i], curmin * nums[i], nums[i])
            res = max(res, curmax)
        return res